AjobK/Athena
# 313 post banner attachment

closes #313 

<br />

 **What is it supposed to do**  
This branch adds functionality for post thumbnails/banners. This is the image the user sees with the post on places like the homepage, and on the top of the page when the user visits the page of a post. The owner of a post can now change the banner by uploading an image to the server. The user is able to do this through the application by selecting or dropping an image. This image can be cropped.

![image](https://user-images.githubusercontent.com/14930277/135699436-b45c9340-a278-4170-9e0b-735a736df080.png)

<br />

 **How can we test your branch**  
Go to one of your posts or create a new one. Hovering over the background thumbnail should show _edit banner_, which can be clicked. This should not happen when you're not logged in as the owner of the post. Only .png or .jpg files should be allowed. Max file size should be 4000 KB. You should only be able to save an image after cropping it. The new image should appear immediately as the new thumbnail, and this should still be there after refreshing. The new image should also be on the homepage (as a thumbnail for the post). Try changing the thumbnail again. Updating the image should delete the old one (from the file system and Attachment table).

<br />

 **Additional notes**  
**You'll have to add some folders to make this branch work**. In _backend/app/public_, add _post/thumbnail_.

Wasn't sure whether I should completely delete the thumbnail after a post is archived or not, so at the moment the thumbnail attachments don't get removed when the post is archived.

When fetching multiple posts, a separate DAO call is executed after for every single post in order to fetch their banner attachments. Maybe there's a more efficient way to do this?
![image](https://user-images.githubusercontent.com/14930277/135699411-eb2d23ca-6c17-47b1-89bb-d4be4fe380dd.png)

<br />

**1. General checklist**

<br />

**2. If relevant, front-end checklist**

<br />

**3. If relevant, back-end checklist**

<br />

**4. If relevant, test these browsers**
